### ISSUE TITLE
False positive redefined-while-unused (F811) for function in __all__

### ISSUE THREAD

**Issue by Tinche** [astral-sh/ruff#8787]
Howdy again.

I'm seeing F811 for a function that I'm exporting via `__all__`.

```python
import sys

__all__ = ["test"]


def test() -> None:
    pass


if sys.version_info() >= (3, 12):

    def test() -> None:
        pass
```

This is a simpified example from cattrs. I think this is a legit use case and shouldn't be flagged.

**Comment by charliermarsh** [comment]
Hmm, on second thought... the easiest solution here might be to instead recommend

```python
import sys

__all__ = ["test"]


if sys.version_info() >= (3, 12):
    def test() -> None:
        pass
else:
    def test() -> None:
        pass
```

If the functions are in different branches, we can correctly identify that they're mutually exclusive. As-is, this will be hard to fix in general.

**Comment by Tinche** [comment]
If it's hard to fix I'm OK with just a `# noqa`. I support Pythons 3.8 - 3.12 and having a branch for each version is just difficult.
